### PR TITLE
gm-connector initial release

### PR DESCRIPTION
**Pull request by greg-szabo**
Closes: #1412

## Description
gm-connector initial implementation with updates to `gm` to support it.

______

For contributor use:

- [x] Added a changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [x] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Comment by adizere** [comment]
Many thanks Greg for broaching this!

@soareschen is the main customer for this feature. I suspect we'll make good use of the gm-connector in bootstrapping and managing chain binaries, in particular for PR #1273.

For context, this PR may be blocked for a few days. Reason is that Soares is sprinting to finish no_std, so it may be a few days until we get his input here. I think that shouldn't be a problem.

**Comment by greg-szabo** [comment]
All improvements are welcome, it was essentially written in a few nights and it's more a line of thought than a crate of code. :) (Option<> instead of Result<>? What was I smoking?) Think of it more as a conversation starter than a PR. (Well, the `gm` part contains a few fixes that are worth merging, but the Rust code is more of a topic of discussion.)

Regarding separate crate: sure, I don't mind either way. Writing a PoC was easier in a separate crate. How much of our e2e testing toolset will be reusable outside of Hermes testing? `gm` was written with Hermes in mind, but the target audience is "anyone who wants to run testnets" and people on the Go side of things picked it up too.

Calling `gaiad` directly: it's up to you. I had the same idea for years (setting up a testnet is like four `gaiad` commands total) but somehow I kept running into people not going through the motions and complaining about how hard it is setting up testnets. `gm` solves issues around configuring these networks: validator key, validator gentx, wallet with available funds, managing endpoints on the host, etc. If you don't need those for testing, we can simply close this ticket without merging. (I'll move the `gm`-related fixes over to a new PR.) As I mentioned, it's more of a topic to discuss.

I'm not sure what you mean by async dispatch of `gm`. The code doesn't exit until `gm` is finished and `gm` goes through the nodes one-by-one and starts them. You don't need to wait after calling the `start()` method, your network is online. At least, that's what I had in mind. I might have missed the target, maybe?

File: scripts/gm/bin/lib-gm [hunk 1]
```diff
@@ -5,9 +5,9 @@ if [ "${DEBUG:-}" = "2" ]; then
 fi
 
 version() {
-  VERSION="v0.0.9"
-  if [ "${OUTPUT:-}" == "json" ]; then
-    printf '{"status": "success", "message": "%s"}' "$VERSION"
+  VERSION="v0.1.0"
```
Comment (adizere) [inline scripts/gm/bin/lib-gm:8]: 0.1 is a relatively big moment! Great work with gm Greg!

**Review by adizere — APPROVED** [review]
Cool work here!

I read through the code & comments and I have two remarks:

1.  It's a bit of a bummer that we are not reusing code between the crates `gm-connector` and `relayer-test` (#1521). We will soon have two crates that accomplish very similar work, i.e., chain scaffolding (or driver or manager).
    - One interpretation of the situation: we probably got here because of the very explicit purpose of `relayer-test` to focus on reproducing relayer logic/core bugs, as opposed of a focus on enabling interoperability with different gaia versions and different chain binaries (regen, akash) which `gm` offers. Given that purpose, I think we probably did the pragmatic decision to add in `relayer-test` support similar to that of `gm-connector`.
    - That being said, it's not a problem at the moment, but in the future I doubt it's a good idea, and it's preferable I believe to have more collaboration and more code reuse. Let me know if I'm misinterpreting anything here.


2. Repository layout:
    - What do you think about moving `gm-connector` out of the top-level and into a `scripts` or `tests` sub-directory? I think the same applies for #1521 BTW (@soareschen).
    - Context: I think at the top level we'll strive to keep either ibc-related artifacts (`module`, `proto`, `relayer`, `relayer-cli`) or simple categories of supporting artifacts (`tests`, `CI`, `docs`) to make navigation simpler. We'll probably refactor everything soon, so let's start experimenting with this scheme already in this PR. WDYT?

File: scripts/gm/bin/lib-gm (hunk 1 above)
Comment (greg-szabo) [reply scripts/gm/bin/lib-gm:8]: Thanks. I wanted to make it count. :)

**Comment by greg-szabo** [comment]
I agree with Adi's concerns, hence I wanted to have more discussion around this.

Regarding folder structure: in tendermint-rs we created a `tools` folder which is essentially its own set of Rust crates completely separated from the main repository. It contains the proto-compiler, some testing frameworks and other tools that required structured upkeep (i.e. `Cargo.toml`) but we didn't want to mix it with the "main" topic of the repository. Check it out: https://github.com/informalsystems/tendermint-rs/tree/master/tools
I would do something similar here and move `gm-connector`, `proto-compiler` and possibly `e2e` into it. I don't consider the scripts folder structured enough for that. (Especially, since it mixes configuration files with unversioned scripts as well as `gm` which is versioned and documented.)

I would like to merge the efforts of `relayer-test` and `gm-connector`, if possible but I could not get any requirements in Berlin. This would also have a big effect on the future of `gm`, so it is crucial to get this in order.

**Comment by adizere** [comment]
Closing in favor of #1606.

This comment https://github.com/informalsystems/ibc-rs/issues/1412#issuecomment-976334390 provides more context on why some of the changes in this PR are not necessary at the moment.

**Pull request closed without merging**